crates/gpui/src/platform/linux/x11/client.rs

```diff
@@ -46,192 +59,217 @@ impl X11Client {
         let xkb_state =
             xkb::x11::state_new_from_device(&xkb_keymap, &xcb_connection, xkb_device_id);
 
-        Self {
-            platform_inner: inner,
-            xcb_connection,
+        let client: Rc<X11Client> = Rc::new(Self {
+            platform_inner: Rc::clone(&inner),
+            xcb_connection: Arc::clone(&xcb_connection),
             x_root_index,
             atoms,
             state: Mutex::new(X11ClientState {
                 windows: HashMap::default(),
+                windows_to_refresh: HashSet::default(),
                 xkb: xkb_state,
             }),
+        });
+
+        let (event_sender, event_receiver) = calloop::channel::channel::<xcb::Event>();
+        {
+            let client = Rc::clone(&client);
+            inner
+                .loop_handle
+                .insert_source(event_receiver, move |event, _, _| match event {
+                    calloop::channel::Event::Msg(event) => {
+                        client.handle_event(event);
+                    }
+                    calloop::channel::Event::Closed => {}
+                })
+                .unwrap();
+        }
+
+        {
+            let xcb_connection = Arc::clone(&xcb_connection);
+            std::thread::spawn(move || loop {
```

Comment:
With this change, we have an extra thread to poll events from the windowing system. It also means that each message has to go through one more step: first get from X11 into this thread, then get from this thread over the channel into the `calloop` event loop, and then be processed.

Reply:
See the comments in https://github.com/Smithay/smithay/blob/a04d9ac8441e19e87a6af6410d0f1ddbf4b293c5/src/utils/x11rb.rs for some reasoning for this choice